### PR TITLE
Http request logging in nuget.exe

### LINKED ISSUE #1324: nuget command line should display http requests it's making to v3 feed when Verbosity is set to detailed
When nuget talks to a v2 feed, it displays http requests when Verbosity is set to true. But it does not show the requests when it connects to a v3 feed.

### PR DESCRIPTION
**Pull request by xavierdecoster**
fixes https://github.com/NuGet/Home/issues/1324

This PR adds request logging in nuget.exe for all requests made by `DownloadResourceV2` and `DownloadResourceV3` when `-verbosity detailed`.

cc @yishaigalatzer @emgarten @feiling 

Sample:

sample

### CODE REVIEW

File: src/NuGet.Clients/NuGet.CommandLine/CommandLineSourceRepositoryProvider.cs [hunk 1]
```diff
@@ -11,20 +11,20 @@ namespace NuGet.CommandLine
...
+        public CommandLineSourceRepositoryProvider(
+            Configuration.IPackageSourceProvider packageSourceProvider,
+            Logging.ILogger logger)
         {
             _packageSourceProvider = packageSourceProvider;
 
-            _resourceProviders = new List<Lazy<INuGetResourceProvider>>();
-            _resourceProviders.AddRange(Protocol.Core.v2.FactoryExtensionsV2.GetCoreV2(Repository.Provider));
-            _resourceProviders.AddRange(Protocol.Core.v3.FactoryExtensionsV2.GetCoreV3(Repository.Provider));
+            _resourceProviders = Repository.Provider.GetCommandline(logger);
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/CommandLineSourceRepositoryProvider.cs]: ToList()

**Comment by xavierdecoster** [comment]
also relates to https://github.com/NuGet/Home/issues/999

**Comment by yishaigalatzer** [comment]
:shipit: when you get another squirrel from someone more intimate with the code than me

**Comment by xavierdecoster** [comment]
ping @emgarten @feiling @MeniZalzman

**Comment by tomkerkhove** [comment]
:shipit:

File: src/NuGet.Clients/NuGet.CommandLine/CommandLineSourceRepositoryProvider.cs [hunk 2]
```diff
@@ -11,20 +12,20 @@ namespace NuGet.CommandLine
     public class CommandLineSourceRepositoryProvider : ISourceRepositoryProvider
     {
         private readonly Configuration.IPackageSourceProvider _packageSourceProvider;
-        private readonly List<Lazy<INuGetResourceProvider>> _resourceProviders;
+        private readonly IEnumerable<Lazy<INuGetResourceProvider>> _resourceProviders;
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/CommandLineSourceRepositoryProvider.cs]: why this change?

File: src/NuGet.Clients/NuGet.CommandLine/ConsoleNuGetResourceProvider.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        public IEnumerable<string> After
+        {
+            get { return _inner.Value.After; }
+        }
+
+        private void OnSendingRequest(object sender, Protocol.Core.Types.WebRequestEventArgs webRequestEventArgs)
+        {
+            _logger.LogDebug($"  {webRequestEventArgs.Method}: {webRequestEventArgs.RequestUri}");
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/ConsoleNuGetResourceProvider.cs]: should this be verbose? Why did we choose debug

File: src/NuGet.Core/NuGet.Protocol.Core.Types/WebRequestEventArgs.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+{
+    public class WebRequestEventArgs : EventArgs
+    {
+        public WebRequestEventArgs(Uri requestUri, string method)
+        {
+            if (requestUri == null)
+            {
+                throw new ArgumentNullException(nameof(requestUri));
+            }
+            if (method == null)
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Protocol.Core.Types/WebRequestEventArgs.cs:14]: newline after }

File: src/NuGet.Core/NuGet.Protocol.Core.Types/WebRequestEventArgs.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+
+namespace NuGet.Protocol.Core.Types
+{
+    public class WebRequestEventArgs : EventArgs
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Protocol.Core.Types/WebRequestEventArgs.cs:5]: Lets avoid event handlers for logging purposes

**Comment by yishaigalatzer** [comment]
change my mind a bit, sorry :watch:

File: src/NuGet.Core/NuGet.Protocol.Core.Types/WebRequestEventArgs.cs (hunk 5 above)
Comment (xavierdecoster) [inline src/NuGet.Core/NuGet.Protocol.Core.Types/WebRequestEventArgs.cs:5]: That's why a started a refactoring branch for logging, as I don't want to pass on the logger through all these layers. Agreed we should get rid of the events for logging.

File: src/NuGet.Clients/NuGet.CommandLine/ConsoleNuGetResourceProvider.cs (hunk 3 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/ConsoleNuGetResourceProvider.cs]: The request was to log these when verbosity = detailed https://github.com/NuGet/Home/issues/1324

File: src/NuGet.Clients/NuGet.CommandLine/CommandLineSourceRepositoryProvider.cs (hunk 2 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/CommandLineSourceRepositoryProvider.cs]: no hard reason, just exposes lesser API as the constructor defines the list, and it should be treated read-only from then on

Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/CommandLineSourceRepositoryProvider.cs]: that's a private member, there is no good reason to do the lesser API thing.

File: src/NuGet.Clients/NuGet.CommandLine/ConsoleNuGetResourceProvider.cs (hunk 3 above)
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/ConsoleNuGetResourceProvider.cs]: and that doesn't work with LogVerbose ? (Don't actually know the answer)

File: src/NuGet.Core/NuGet.Protocol.Core.Types/WebRequestEventArgs.cs (hunk 5 above)
Comment (xavierdecoster) [inline src/NuGet.Core/NuGet.Protocol.Core.Types/WebRequestEventArgs.cs:5]: Note that I just moved the existing `WebRequestEventArgs` down into a common layer. The Package Manager Console was already using it to log stuff using event handlers.

Created an alternative PR that only affects nuget.exe code without the need for a rather impactful refactoring of the logging code: https://github.com/NuGet/NuGet.Client/issues/138

**Comment by xavierdecoster** [comment]
Replacing this PR with https://github.com/NuGet/NuGet.Client/issues/138 (least impacting changes)
